File: apps/agent-daemon/src/lib/update.ts
import { realpathSync } from 'node:fs';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { homedir, platform } from 'node:os';
import { dirname, join } from 'node:path';

import { loadUpdateEnvConfig } from '../config.js';

/**
 * `npm install -g @themoltnet/agent-daemon@latest` resolves the registry's own
 * dist-tag, so for an npm install the registry is the precise answer — the
 * manifest serves the themolt.net pin, which gates the bundle installer's
 * default and says nothing about what npm already has. It is also the origin an
 * npm install already talks to, so this adds no new dependency in CI, where the
 * daemon runs as `npx @themoltnet/agent-daemon` and would otherwise reach for
 * api.github.com from shared runner IPs.
 *
 * Bundle and direct installs ask the release listing, and their upgrade command
 * passes `MOLTNET_AGENT_VERSION=latest` so the installer resolves the same thing
 * rather than falling back to its own pinned default. Reporting the pin here
 * would hide releases that are already downloadable; reporting the release
 * without the sentinel would advertise a version the command could not deliver.
 * The pin still governs a fresh `curl .../install/agent | sh`.
 */
export const UPDATE_RELEASES_URL =
  'https://api.github.com/repos/getlarge/themoltnet/releases?per_page=100';
const RELEASE_TAG_PREFIX = 'agent-daemon-v';
export const UPDATE_NPM_REGISTRY_URL =
  'https://registry.npmjs.org/@themoltnet/agent-daemon/latest';
export const UPDATE_CACHE_TTL_MS = 24 * 60 * 60 * 1000;
const UPDATE_ERROR_CACHE_TTL_MS = 5 * 60 * 1000;
export type DaemonInstallMethod = 'bundle' | 'npm' | 'direct';
export interface UpdateResult {
  currentVersion: string;
  latestVersion?: string;
  updateAvailable: boolean;
  installMethod: DaemonInstallMethod;
  releaseUrl: string;
  command: string;
}
interface UpdateCache {
  checkedAt: string;
  latest?: string;
  error?: string;
}

/**
 * Node does not resolve `process.argv[1]`: invoked through npm's bin shim it
 * reports `/usr/local/bin/moltnet-agent`, not the `node_modules` path this
 * matches on. Every global npm install therefore looked like `direct` and was
 * offered the bundle installer, which would drop a bundle on top of an npm
 * install. Resolve first; a missing path falls back to the input.
 */
export function resolveDaemonExecutable(executable: string): string {
  if (!executable) return executable;
  try {
    return realpathSync(executable);
  } catch {
    return executable;
  }
}

export function detectDaemonInstallMethod(
  executable = process.argv[1] ?? '',
): DaemonInstallMethod {
  const path = resolveDaemonExecutable(executable).replaceAll('\\', '/');
  if (path.includes('/node_modules/@themoltnet/agent-daemon/')) return 'npm';
  if (path.includes('/.local/share/moltnet/') || path.includes('/opt/moltnet/'))
    return 'bundle';
  return 'direct';
}
export function daemonUpdateCommand(method: DaemonInstallMethod): string {
  if (method === 'npm') return 'npm install -g @themoltnet/agent-daemon@latest';
  // Without the sentinel the installer would fall back to the version it was
  // pinned with, reinstalling what the user already has while the check keeps
  // reporting an update. `latest` re-resolves on every run, so the command
  // stays a stable string rather than embedding a number that goes stale.
  return 'curl -fsSL https://themolt.net/install/agent | MOLTNET_AGENT_VERSION=latest sh';
}
export async function checkDaemonUpdate(input: {
  currentVersion: string;
  force?: boolean;
  executable?: string;
  fetchFn?: typeof fetch;
  now?: Date;
}): Promise<UpdateResult> {
  const executable = input.executable ?? process.argv[1] ?? 'moltnet-agent';
  const installMethod = detectDaemonInstallMethod(executable);
  const result: UpdateResult = {
    currentVersion: normalizeVersion(input.currentVersion),
    updateAvailable: false,
    installMethod,
    releaseUrl: 'https://themolt.net/download',
    command: daemonUpdateCommand(installMethod),
  };
  const now = input.now ?? new Date();
  if (!input.force) {
    const cached = await readCache();
    const ttl = cached?.latest
      ? UPDATE_CACHE_TTL_MS
      : UPDATE_ERROR_CACHE_TTL_MS;
    if (cached && now.getTime() - Date.parse(cached.checkedAt) < ttl) {
      result.latestVersion = cached.latest;
      result.updateAvailable =
        compareVersions(cached.latest, result.currentVersion) > 0;
      return result;
    }
  }
  const viaNpm = installMethod === 'npm';
  const source = viaNpm ? UPDATE_NPM_REGISTRY_URL : UPDATE_RELEASES_URL;
  const label = viaNpm ? 'npm registry' : 'release listing';
  try {
    const response = await (input.fetchFn ?? fetch)(source, {
      signal: AbortSignal.timeout(5000),
    });
    if (!response.ok)
      throw new Error(`${label} returned HTTP ${response.status}`);
    const body: unknown = await response.json();
    const latest = viaNpm ? distTagVersion(body) : newestReleaseVersion(body);
    if (!latest) throw new Error(`${label} has no valid agent version`);
    await writeCache({ checkedAt: now.toISOString(), latest }).catch(
      () => undefined,
    );
    result.latestVersion = latest;
    result.updateAvailable = compareVersions(latest, result.currentVersion) > 0;
    return result;
  } catch (error) {
    await writeCache({
      checkedAt: now.toISOString(),
      error: error instanceof Error ? error.message : String(error),
    }).catch(() => undefined);
    throw new Error(
      `could not check for MoltNet agent updates: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
}
function distTagVersion(value: unknown): string | undefined {
  if (!value || typeof value !== 'object') return undefined;
  const version = (value as { version?: unknown }).version;
  return typeof version === 'string' && validVersion(version)
    ? normalizeVersion(version)
    : undefined;
}
/**
 * The listing is ordered by creation date, not version, so compare every
 * candidate rather than trusting position. Drafts are filtered explicitly:
 * unauthenticated callers never see them, but a token would, and this
 * repository carries stuck drafts that would otherwise look newest.
 */
function newestReleaseVersion(value: unknown): string | undefined {
  if (!Array.isArray(value)) return undefined;
  let newest: string | undefined;
  for (const entry of value) {
    if (!entry || typeof entry !== 'object') continue;
    const {
      tag_name: tag,
      draft,
      prerelease,
    } = entry as Record<string, unknown>;
    if (draft === true || prerelease === true) continue;
    if (typeof tag !== 'string' || !tag.startsWith(RELEASE_TAG_PREFIX))
      continue;
    const candidate = tag.slice(RELEASE_TAG_PREFIX.length);
    if (!validVersion(candidate)) continue;
    if (!newest || compareVersions(candidate, newest) > 0)
      newest = normalizeVersion(candidate);
  }
  return newest;
}
function normalizeVersion(value: string): string {
  return value.trim().replace(/^v/, '');
}
function validVersion(value: string): boolean {
  return /^v?\d+\.\d+\.\d+$/.test(value);
}
export function compareVersions(a: string | undefined, b: string): number {
  if (!a || !validVersion(a) || !validVersion(b)) return 0;
  const left = normalizeVersion(a).split('.').map(Number);
  const right = normalizeVersion(b).split('.').map(Number);
  for (let i = 0; i < 3; i += 1) {
    const leftPart = left[i] ?? 0;
    const rightPart = right[i] ?? 0;
    if (leftPart > rightPart) return 1;
    if (leftPart < rightPart) return -1;
  }
  return 0;
}
function cachePath(): string {
  const env = loadUpdateEnvConfig();
  return join(
    env.xdgCacheHome ||
      (platform() === 'win32'
        ? env.localAppData || join(homedir(), 'AppData', 'Local')
        : join(homedir(), '.cache')),
    'moltnet',
    'updates',
    'agent.json',
  );
}
async function readCache(): Promise<UpdateCache | undefined> {
  try {
    return JSON.parse(await readFile(cachePath(), 'utf8')) as UpdateCache;
  } catch {
    return undefined;
  }
}
async function writeCache(cache: UpdateCache): Promise<void> {
  const path = cachePath();
  await mkdir(dirname(path), { recursive: true, mode: 0o700 });
  await writeFile(path, JSON.stringify(cache), { mode: 0o600 });
}

File: apps/moltnet-cli/update.go
package main

// Stable release update discovery intentionally has no relationship with the
// API client or credentials.  In particular, `update check` must remain useful
// on an uninitialised machine and safe to run from CI.

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/spf13/cobra"
)

// Update discovery asks GitHub which cli releases exist, not
// themolt.net/download/manifest.json.  That manifest serves the deliberately
// pinned version behind /install and /download/*, which is a reviewed decision
// about what the website hands a brand-new user — not a statement about what
// has shipped.  Homebrew, apt, scoop and npm all receive a release the moment
// the release job finishes, so sourcing the *notification* from the pin only
// hid versions users could already install, for as long as the pin PR sat
// unmerged.
const updateReleasesURL = "https://api.github.com/repos/getlarge/themoltnet/releases?per_page=100"
const updateCacheTTL = 24 * time.Hour
const updateTagPrefix = "cli-v"

// Package variables make the transport boundary testable without changing the
// public endpoint used by released binaries.
var cliUpdateReleasesURL = updateReleasesURL
var cliUpdateHTTPClient = &http.Client{}

type githubRelease struct {
	TagName    string `json:"tag_name"`
	Draft      bool   `json:"draft"`
	Prerelease bool   `json:"prerelease"`
}

type updateCache struct {
	CheckedAt time.Time `json:"checkedAt"`
	Latest    string    `json:"latest,omitempty"`
	Error     string    `json:"error,omitempty"`
}

type updateResult struct {
	Current         string `json:"currentVersion"`
	Latest          string `json:"latestVersion,omitempty"`
	UpdateAvailable bool   `json:"updateAvailable"`
	InstallMethod   string `json:"installMethod"`
	ReleaseURL      string `json:"releaseUrl"`
	Command         string `json:"command"`
}

func newUpdateCmd(version string) *cobra.Command {
	var jsonOutput bool
	check := &cobra.Command{
		Use:   "check",
		Short: "Check the stable MoltNet CLI release",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			result, err := checkCLIUpdate(cmd.Context(), version, true)
			if err != nil {
				return err
			}
			if jsonOutput {
				return json.NewEncoder(cmd.OutOrStdout()).Encode(result)
			}
			if result.UpdateAvailable {
				fmt.Fprintf(cmd.OutOrStdout(), "MoltNet CLI %s is available (you have %s). Run: %s\\n", result.Latest, result.Current, result.Command)
			} else {
				fmt.Fprintf(cmd.OutOrStdout(), "MoltNet CLI %s is up to date.\\n", result.Current)
			}
			return nil
		},
	}
	check.Flags().BoolVar(&jsonOutput, "json", false, "Print machine-readable JSON")
	parent := &cobra.Command{Use: "update", Short: "Inspect MoltNet CLI updates", Args: cobra.NoArgs, RunE: func(cmd *cobra.Command, args []string) error { return cmd.Help() }}
	parent.AddCommand(check)
	return parent
}

func checkCLIUpdate(ctx context.Context, current string, force bool) (updateResult, error) {
	exe, _ := os.Executable()
	exe = resolveCLIExecutable(exe)
	method := detectCLIInstallMethod(exe)
	result := updateResult{Current: normalVersion(current), InstallMethod: method, ReleaseURL: "https://themolt.net/download", Command: cliUpdateCommand(method, exe)}
	cache, cacheErr := readUpdateCache("cli")
	if !force && cacheErr == nil && time.Since(cache.CheckedAt) < updateCacheTTL {
		result.Latest = cache.Latest
		result.UpdateAvailable = compareVersions(result.Latest, result.Current) > 0
		return result, nil
	}
	latest, err := fetchCLILatest(ctx)
	if err != nil {
		_ = writeUpdateCache("cli", updateCache{CheckedAt: time.Now().UTC(), Error: err.Error()})
		return result, fmt.Errorf("could not check for MoltNet CLI updates: %w", err)
	}
	_ = writeUpdateCache("cli", updateCache{CheckedAt: time.Now().UTC(), Latest: latest})
	result.Latest = latest
	result.UpdateAvailable = compareVersions(latest, result.Current) > 0
	return result, nil
}

func fetchCLILatest(ctx context.Context) (string, error) {
	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, cliUpdateReleasesURL, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("User-Agent", "moltnet-cli")
	res, err := cliUpdateHTTPClient.Do(req)
	if err != nil {
		return "", err
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		return "", fmt.Errorf("release listing returned HTTP %d", res.StatusCode)
	}
	var releases []githubRelease
	if err := json.NewDecoder(res.Body).Decode(&releases); err != nil {
		return "", fmt.Errorf("invalid release listing: %w", err)
	}
	latest := ""
	for _, release := range releases {
		// An unauthenticated caller never sees drafts, but a token in CI does,
		// and this repository carries stuck drafts months old — one would
		// otherwise masquerade as the newest release.  Same lesson as
		// tools/ci/check-download-pins.sh.
		if release.Draft || release.Prerelease {
			continue
		}
		if !strings.HasPrefix(release.TagName, updateTagPrefix) {
			continue
		}
		candidate := strings.TrimPrefix(release.TagName, updateTagPrefix)
		if !validVersion(candidate) {
			continue
		}
		// The listing is ordered by creation date, not version: a patch cut on
		// an older line can appear ahead of a newer minor.  Compare every
		// candidate rather than trusting position.
		if latest == "" || compareVersions(candidate, latest) > 0 {
			latest = candidate
		}
	}
	if latest == "" {
		return "", errors.New("no valid CLI release found")
	}
	return latest, nil
}

func updateCachePath(product string) (string, error) {
	d, err := os.UserCacheDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(d, "moltnet", "updates", product+".json"), nil
}
func readUpdateCache(product string) (updateCache, error) {
	var c updateCache
	p, err := updateCachePath(product)
	if err != nil {
		return c, err
	}
	b, err := os.ReadFile(p)
	if err != nil {
		return c, err
	}
	err = json.Unmarshal(b, &c)
	return c, err
}
func writeUpdateCache(product string, c updateCache) error {
	p, err := updateCachePath(product)
	if err != nil {
		return err
	}
	if err = os.MkdirAll(filepath.Dir(p), 0700); err != nil {
		return err
	}
	b, err := json.Marshal(c)
	if err != nil {
		return err
	}
	return os.WriteFile(p, b, 0600)
}

func normalVersion(v string) string { return strings.TrimPrefix(strings.TrimSpace(v), "v") }
func validVersion(v string) bool {
	p := strings.Split(normalVersion(v), ".")
	if len(p) != 3 {
		return false
	}
	for _, s := range p {
		if s == "" {
			return false
		}
		for _, r := range s {
			if r < '0' || r > '9' {
				return false
			}
		}
	}
	return true
}
func compareVersions(a, b string) int {
	if !validVersion(a) || !validVersion(b) {
		return 0
	}
	var ai, bi [3]int
	fmt.Sscanf(normalVersion(a), "%d.%d.%d", &ai[0], &ai[1], &ai[2])
	fmt.Sscanf(normalVersion(b), "%d.%d.%d", &bi[0], &bi[1], &bi[2])
	for i := range ai {
		if ai[i] > bi[i] {
			return 1
		}
		if ai[i] < bi[i] {
			return -1
		}
	}
	return 0
}

// os.Executable reports the path the binary was invoked through, which for a
// Homebrew cask on macOS is the shim (/opt/homebrew/bin/moltnet), not its
// Caskroom target.  Install-method detection matches on the real location, so
// every macOS brew install was misreported as "direct" and offered the curl
// installer — which would overwrite Homebrew's symlink and desync its version
// tracking.  Linux was unaffected: os.Executable reads /proc/self/exe, which is
// already resolved.
func resolveCLIExecutable(exe string) string {
	if exe == "" {
		return exe
	}
	resolved, err := filepath.EvalSymlinks(exe)
	if err != nil {
		return exe
	}
	return resolved
}

func detectCLIInstallMethod(exe string) string {
	p := filepath.ToSlash(exe)
	switch {
	case strings.Contains(p, "/Caskroom/moltnet/"):
		return "homebrew"
	case strings.Contains(p, "/scoop/apps/moltnet/"):
		return "scoop"
	case strings.Contains(p, "/node_modules/@themoltnet/cli/"):
		return "npm"
	case runtime.GOOS == "linux" && aptOwnsMoltnet(exe):
		return "apt"
	case strings.Contains(p, "/.local/share/moltnet/") || strings.Contains(p, "/opt/moltnet/"):
		return "standalone"
	default:
		return "direct"
	}
}
func aptOwnsMoltnet(exe string) bool {
	b, err := os.ReadFile("/var/lib/dpkg/info/moltnet.list")
	return err == nil && strings.Contains(string(b), exe)
}

func isCLIWorkspaceInvocation() bool {
	exe, err := os.Executable()
	if err != nil {
		return false
	}
	return strings.Contains(filepath.ToSlash(exe), "/apps/moltnet-cli/")
}
func cliUpdateCommand(method, exe string) string {
	switch method {
	case "homebrew":
		// `brew upgrade` consults the local tap clone; third-party taps are not
		// in the core JSON API.  Without a fetch, a stale clone still advertises
		// the installed version and the upgrade no-ops with "already installed".
		return "brew update && brew upgrade --cask moltnet"
	case "apt":
		return "sudo apt update && sudo apt install --only-upgrade moltnet"
	case "scoop":
		// scoop-update.ps1 calls Sync-Bucket only when invoked with no app
		// argument, so `scoop update moltnet` alone never refreshes the bucket.
		return "scoop update && scoop update moltnet"
	case "npm":
		return "npm install -g @themoltnet/cli@latest"
	default:
		return fmt.Sprintf("curl -fsSL https://themolt.net/install/cli | sh -s -- --replace %q", exe)
	}
}
